python/dgl/backend/tensorflow/tensor.py

```diff
@@ -0,0 +1,564 @@
+    return spmat, None
+
+
+def sparse_matrix_indices(spmat):
+    return ('coo', spmat.indices)
+
+
+def is_tensor(obj):
+    return isinstance(obj, tf.Tensor)
+
+
+def shape(input):
+    return input.shape
+
+
+def dtype(input):
+    return input.dtype
+
+
+def ndim(input):
+    return input.ndim
+
+
+def context(input):
+    return input.device
+
+
+def device_type(ctx):
+    return tf.DeviceSpec.from_string(ctx).device_type.lower()
+
+
+def device_id(ctx):
+    return tf.DeviceSpec.from_string(ctx).device_index
+
+
+def astype(input, ty):
+    return tf.cast(input, dtype=ty)
+
+
+def asnumpy(input):
+    if isinstance(input, tf.SparseTensor):
+        # tf.sparse.to_dense assume sorted indices, need to turn off validate_indices in our cases
+        return tf.sparse.to_dense(input, validate_indices=False).numpy()
+    else:
+        return input.numpy()
+
+
+def copy_to(input, ctx):
+    with tf.device(ctx):
+        new_tensor = tf.identity(input)
+    return new_tensor
+
+
+def sum(input, dim, keepdims=False):
+    return tf.reduce_sum(input, axis=dim, keepdims=keepdims)
+
+
+def reduce_sum(input):
+    return tf.reduce_sum(input)
+
+
+def mean(input, dim):
+    return tf.reduce_mean(input, axis=dim)
+
+
+def reduce_mean(input):
+    return th.reduce_mean(input)
+
+
+def max(input, dim):
+    return tf.reduce_max(input, axis=dim)
+
+
+def reduce_max(input):
+    return tf.reduce_max(input)
+
+
+def min(input, dim):
+    return tf.reduce_min(input, axis=dim)
+
+
+def reduce_min(input):
+    return tf.reduce_min(input)
+
+
+def argsort(input, dim, descending):
+    if descending:
+        return tf.cast(tf.argsort(input, axis=dim, direction="DESCENDING"), dtype=tf.int64)
+    else:
+        return tf.cast(tf.argsort(input, axis=dim, direction="ASCENDING"), dtype=tf.int64)
+
+
+def topk(input, k, dim, descending=True):
+    # tf's topk doesn't support direction and dimension, use sort instead
+    direction = 'DESCENDING' if descending else 'ASCENDING'
+    return slice_axis(tf.sort(input, axis=dim, direction=direction), dim, 0, k)
+
+
+def argtopk(input, k, dim, descending=True):
+    # tf's topk doesn't support direction and dimension, use argsort instead
+    direction = 'DESCENDING' if descending else 'ASCENDING'
+    return slice_axis(tf.argsort(input, axis=dim, direction=direction), dim, 0, k)
+
+
+def exp(input):
+    return tf.exp(input)
+
+
+def softmax(input, dim=-1):
+    return tf.math.softmax(input, axis=dim)
+
+
+def cat(seq, dim):
+    return tf.concat(seq, axis=dim)
+
+
+def stack(seq, dim):
+    return tf.stack(seq, axis=dim)
+
+
+def split(input, sizes_or_sections, dim):
+    return tf.split(input, sizes_or_sections, axis=dim)
+
+
+def repeat(input, repeats, dim):
+    tile_shape = np.ones(input.ndim, dtype=np.int32)
+    tile_shape[dim] = repeats
+    return tf.tile(input, tf.constant(tile_shape))
+
+
+def gather_row(data, row_index):
+    # Weird implementation, because tf.gather require one dimension to be batch_dim
```

Comment:
What is the specific problem here?

Reply:
No problem here. I wrote this when I was not familiar with tf API.

Reply:
No problem here. I wrote this when I was not familiar with tf API.